### PR TITLE
[ML] Allow put trained model to overwrite pre-packaged models

### PR DESCRIPTION
**Pull request by jonathan-buttner**
This PR allows users to rerun the put trained model API request for pre-packaged models. If the model is a pre-packaged one and already downloaded it'll simply overwrite it.

The approach I took was to switch to using the [index operation instead of create](https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-index_.html#docs-index-api-query-params) if the model is pre-packaged. To determine if the model is pre-packaged, we call the `TrainedModelConfig::isPackageModel` within the `TransportPutTrainedModelAction` class. The `index` operation will overwrite an existing document if it exists, where as `create` will throw an error if the document id already exists.

This PR also adds logic to the put trained model API so that it checks to see if a download task already exists. If one does, it will either wait for the task to complete (if the `wait_for_completion=true` flag is set) or immediately returns the model information. If no task exists it proceeds with storing the model.

## Testing

To test that overwriting works, load the elser model multiple times without deleting it

```
curl --location --request PUT 'http://localhost:9200/_ml/trained_models/.elser_model_1?wait_for_completion=true' \
--header 'Content-Type: application/json' \
--header 'Authorization: Basic ZWxhc3RpYzpwYXNzd29yZA==' \
--data '{"input":{"field_names":["text_field"]}}'
```

To test that overwriting does not work for other models, attempt to load a custom model multiple times without deleting. To do this I modified eland a bit:

<details><summary>eland diff</summary>

```
diff --git a/bin/eland_import_hub_model b/bin/eland_import_hub_model
index 94ec4d0..31b8f12 100755
--- a/bin/eland_import_hub_model
+++ b/bin/eland_import_hub_model
@@ -250,26 +250,26 @@ def main():
         ptm = PyTorchModel(
             es, args.es_model_id if args.es_model_id else tm.elasticsearch_model_id()
         )
-        model_exists = (
-            es.options(ignore_status=404)
-            .ml.get_trained_models(model_id=ptm.model_id)
-            .meta.status
-            == 200
-        )
-
-        if model_exists:
-            if args.clear_previous:
-                logger.info(f"Stopping deployment for model with id '{ptm.model_id}'")
-                ptm.stop()
-
-                logger.info(f"Deleting model with id '{ptm.model_id}'")
-                ptm.delete()
-            else:
-                logger.error(f"Trained model with id '{ptm.model_id}' already exists")
-                logger.info(
-                    "Run the script with the '--clear-previous' flag if you want to overwrite the existing model."
-                )
-                exit(1)
+        # model_exists = (
+        #     es.options(ignore_status=404)
+        #     .ml.get_trained_models(model_id=ptm.model_id)
+        #     .meta.status
+        #     == 200
+        # )
+
+        # if model_exists:
+        #     if args.clear_previous:
+        #         logger.info(f"Stopping deployment for model with id '{ptm.model_id}'")
+        #         ptm.stop()
+        #
+        #         logger.info(f"Deleting model with id '{ptm.model_id}'")
+        #         ptm.delete()
+        #     else:
+        #         logger.error(f"Trained model with id '{ptm.model_id}' already exists")
+        #         logger.info(
+        #             "Run the script with the '--clear-previous' flag if you want to overwrite the existing model."
+        #         )
+        #         exit(1)

         logger.info(f"Creating model with id '{ptm.model_id}'")
         ptm.put_config(config=config)
```

</details>

Rebuild the docker image:

```
docker build -t elastic/eland .
```

And import the model twice, you should get an error

```
docker run --network host -it --rm elastic/eland \
    eland_import_hub_model \
      --url http://host.docker.internal:9200 \
      --hub-model-id elastic/distilbert-base-uncased-finetuned-conll03-english \
      --task-type ner \
      --start \
      -u elastic \
      -p password \
      --insecure
```

Error

```
2023-07-18 18:50:04,261 INFO : Creating model with id 'elastic__distilbert-base-uncased-finetuned-conll03-english'
Traceback (most recent call last):
  File "/usr/local/bin/eland_import_hub_model", line 292, in <module>
    main()
  File "/usr/local/bin/eland_import_hub_model", line 275, in main
    ptm.put_config(config=config)
  File "/usr/local/lib/python3.9/dist-packages/eland/ml/pytorch/_pytorch_model.py", line 78, in put_config
    self._client.ml.put_trained_model(model_id=self.model_id, **config_map)
  File "/usr/local/lib/python3.9/dist-packages/elasticsearch/_sync/client/utils.py", line 414, in wrapped
    return api(*args, **kwargs)
  File "/usr/local/lib/python3.9/dist-packages/elasticsearch/_sync/client/ml.py", line 3301, in put_trained_model
    return self.perform_request(  # type: ignore[return-value]
  File "/usr/local/lib/python3.9/dist-packages/elasticsearch/_sync/client/_base.py", line 389, in perform_request
    return self._client.perform_request(
  File "/usr/local/lib/python3.9/dist-packages/elasticsearch/_sync/client/_base.py", line 320, in perform_request
    raise HTTP_EXCEPTIONS.get(meta.status, ApiError)(
elasticsearch.BadRequestError: BadRequestError(400, 'status_exception', 'Cannot create model [elastic__distilbert-base-uncased-finetuned-conll03-english] the id is the same as an current model deployment')
```

### CODE REVIEW

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/persistence/IndexRequestCreator.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+import java.util.Set;
+
+public class IndexRequestCreator {
+    private static final ToXContent.Params FOR_INTERNAL_STORAGE_PARAMS = new ToXContent.MapParams(
+        Collections.singletonMap(ToXContentParams.FOR_INTERNAL_STORAGE, "true")
+    );
+
+    private static final Set<String> PRE_PACKAGED_MODELS = Set.of(".elser_model_1");
+
+    public static IndexRequest create(String modelId, String docId, String index, ToXContentObject body) {
```
Comment (jonathan-buttner) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/persistence/IndexRequestCreator.java:30]: Another idea I had here was this method to accept an interface that looks something like:

```
interface Identifiable {
  String getModelId();
  String getDocId();
}
```

I wasn't sure if that was overkill or didn't make sense to have `getDocId()` always defined for classes like `TrainedModelConfig`. I suppose I could extend the class to include that functionality.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/persistence/IndexRequestCreator.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+ */
+public class IndexRequestCreator {
+    private static final ToXContent.Params FOR_INTERNAL_STORAGE_PARAMS = new ToXContent.MapParams(
+        Collections.singletonMap(ToXContentParams.FOR_INTERNAL_STORAGE, "true")
+    );
+
+    /**
+     * This is package private so the tests can access it
+     */
+    static final Set<String> PRE_PACKAGED_MODELS = Set.of(".elser_model_1");
```
Comment (jonathan-buttner) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/persistence/IndexRequestCreator.java:35]: This strings `.elser_model_1` also exists at: https://github.com/jonathan-buttner/elasticsearch/blob/ml-put-trained-allow-overwrite/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/StartTrainedModelDeploymentAction.java#L658

Any thoughts on a good central place to put this?

**Comment by jonathan-buttner** [comment]
@elasticmachine run elastisearch-ci/part-3

**Comment by jonathan-buttner** [comment]
@elasticsearchmachine run elastisearch-ci/part-3

**Comment by jonathan-buttner** [comment]
@elasticsearchmachine test this please

**Comment by jonathan-buttner** [comment]
@elasticsearchmachine update branch

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/persistence/IndexRequestCreator.java (hunk 2 above)
Comment (hendrikmuhs) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/persistence/IndexRequestCreator.java:35]: hard coding in code has a big disadvantage: We will never be able to push out a new model after a release is out. It is also extra burden to maintain this list and keep it in sync. I think we don't want to go this way.

The `isElserModel` method has been implemented as a short term solution, because we struggled with estimating the required memory usage. It turned out it is different to the estimation we have for other pytorch models, but we wanted to play safe, that's why we implemented it this way.

Long - hopefully not so long - term the plan is to replace the estimate and let the model define its memory usage as part of the metadata.

Long story short: I think the decision here should not be based on string matching, but simply whether it is a pre-packaged model or not.

Comment (hendrikmuhs) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/persistence/IndexRequestCreator.java:35]: I suggest to add a parameter to `PutTrainedModelDefinitionPartAction.Request`, e.g. `overwrite` with a default of `false`. `ModelImporter` sets this to `true`. If `true` you use the index not the create option.

Note: There is no immediate need to expose `overwrite` in the REST layer, we can keep this internal.

Comment (jonathan-buttner) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/persistence/IndexRequestCreator.java:35]: Thanks for the background and that makes sense. Just to make sure I understand, you're suggesting instead of having the the `IndexRequestCreator` have a `Set` to determine whether it is a prepackaged model, we have the `ModelImporter` send a flag indicating whether we should overwrite and pass that down through to the `TrainedModelProvider` which will determine whether we use `Index` or `Create`?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/PutTrainedModelDefinitionPartAction.java [hunk 3]
```diff
@@ -78,6 +91,11 @@ public Request(StreamInput in) throws IOException {
             this.part = in.readVInt();
             this.totalDefinitionLength = in.readVLong();
             this.totalParts = in.readVInt();
+            if (in.getTransportVersion().onOrAfter(TransportVersion.V_8_500_038)) {
```
Comment (jonathan-buttner) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/PutTrainedModelDefinitionPartAction.java:94]: I _think_ we need this version check right?

File: x-pack/plugin/ml-package-loader/src/main/java/org/elasticsearch/xpack/ml/packageloader/action/TransportLoadTrainedModelPackage.java [hunk 4]
```diff
@@ -190,7 +190,7 @@ public TaskId getParentTask() {
 
             @Override
             public CancellableTask createTask(long id, String type, String action, TaskId parentTaskId, Map<String, String> headers) {
-                return new CancellableTask(id, type, action, getDescription(), parentTaskId, headers);
+                return new CancellableTask(id, type, action, "modelId=" + request.getModelId(), parentTaskId, headers);
```
Comment (jonathan-buttner) [inline x-pack/plugin/ml-package-loader/src/main/java/org/elasticsearch/xpack/ml/packageloader/action/TransportLoadTrainedModelPackage.java:193]: This is needed so we can search for a particular task. We could also put this in the `action` field but when I looked around it didn't seem like that was a common place to do this.

Comment (jonathan-buttner) [reply x-pack/plugin/ml-package-loader/src/main/java/org/elasticsearch/xpack/ml/packageloader/action/TransportLoadTrainedModelPackage.java:193]: TODO: move this to a common place so that the request knows how to build it

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportPutTrainedModelAction.java [hunk 5]
```diff
@@ -235,23 +238,26 @@ protected void masterOperation(
...
+                        downloadTriggered -> listener.onResponse(new PutTrainedModelAction.Response(configToReturn)),
+                        listener::onFailure
+                    )
+                );
+            } else {
+                listener.onResponse(new PutTrainedModelAction.Response(configToReturn));
+            }
+        }, listener::onFailure);
+
+        var isPackageModel = config.isPackagedModel();
```
Comment (jonathan-buttner) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportPutTrainedModelAction.java:258]: We need to leverage `isPackageModel` to determine if we can overwrite when we call `storeTrainedModel` below

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/persistence/TrainedModelProvider.java [hunk 6]
```diff
@@ -306,6 +334,14 @@ public void storeTrainedModelDefinitionDoc(
     }
 
     public void storeTrainedModelMetadata(TrainedModelMetadata trainedModelMetadata, ActionListener<Void> listener) {
+        storeTrainedModelMetadata(trainedModelMetadata, listener, false);
+    }
+
+    public void storeTrainedModelMetadata(
```
Comment (jonathan-buttner) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/persistence/TrainedModelProvider.java:340]: This change isn't really necessary because no one outside of the function above calls this method. I can revert the change but I figured it it'd make it consistent with the other `store*` entry points.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportPutTrainedModelAction.java [hunk 7]
```diff
@@ -304,22 +308,106 @@ protected void masterOperation(
...
+            if (models.getResources().results().size() == 0) {
+                listener.onFailure(
+                    new ElasticsearchStatusException(
+                        "No model information found for a concurrent create model execution for model id [{}]",
+                        RestStatus.INTERNAL_SERVER_ERROR,
+                        modelId
+                    )
+                );
+            } else {
+                listener.onResponse(new PutTrainedModelAction.Response(models.getResources().results().get(0)));
```
Comment (jonathan-buttner) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportPutTrainedModelAction.java:399]: In the event that a task exists this will return the model information for the original request. So if the second request that hit this code path is attempting to change say the tags, it will be ignored. Do we think that's acceptable?

An alternative would be to return a failure saying that a download is occurring and it needs to be cancelled before we can attempt downloading again?

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/persistence/IndexRequestCreator.java (hunk 2 above)
Comment (hendrikmuhs) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/persistence/IndexRequestCreator.java]: yes, if it originates from `ModelImporter` we know it originates from a pre-packaged model.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportPutTrainedModelAction.java (hunk 7 above)
Comment (hendrikmuhs) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportPutTrainedModelAction.java:399]: Pre-packaged models don't allow changing the description, tags and other fields.

I wonder: can we distinguish between different downloads using the task API?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/PutTrainedModelDefinitionPartAction.java (hunk 3 above)
Comment (droberts195) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/PutTrainedModelDefinitionPartAction.java:94]: We need a new version constant (which would currently be `V_8_500_040`). There's a comment in `TransportVersion.java` about how to add a new one.

If `V_8_500_038` were used here then serverless upgrades would fail, as already-deployed nodes in serverless will think they understand everything included in `V_8_500_038`.

Comment (jonathan-buttner) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/PutTrainedModelDefinitionPartAction.java:94]: Thanks Dave! Yeah I forgot to merge upstream so I added 38 but someone had already added it. I'll add a new one 👍

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportPutTrainedModelAction.java (hunk 7 above)
Comment (jonathan-buttner) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportPutTrainedModelAction.java:399]: > Pre-packaged models don't allow changing the description, tags and other fields.

Oh ok great.

> I wonder: can we distinguish between different downloads using the task API?

Yeah I believe that currently works, the logic I added only looks for tasks with a specific model id.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/MlTasks.java [hunk 8]
```diff
@@ -51,6 +51,7 @@ public final class MlTasks {
     public static final String DATAFEED_TASK_ID_PREFIX = "datafeed-";
     public static final String DATA_FRAME_ANALYTICS_TASK_ID_PREFIX = "data_frame_analytics-";
     public static final String JOB_SNAPSHOT_UPGRADE_TASK_ID_PREFIX = "job-snapshot-upgrade-";
+    private static final String DOWNLOAD_MODEL_TASK_DESCRIPTION_PREFIX = "modelId=";
```
Comment (hendrikmuhs) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/MlTasks.java:54]: nit, all the other prefixes use `-` and snake case, what about `model_id-`?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/MlTasks.java [hunk 9]
```diff
@@ -477,4 +478,14 @@ public static String prettyPrintTaskName(String taskName) {
...
         };
     }
+
+    /**
+     * Builds the task description from the model id that initiated the task.
+     *
+     * @param modelId a string that identifies the model
+     * @return a string representing the task description
+     */
+    public static String createTaskDescription(String modelId) {
```
Comment (hendrikmuhs) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/MlTasks.java:488]: nit: I think `downloadModelTaskDescription` would fit better with the other naming choices in this class

**Review by hendrikmuhs — APPROVED** [review]
LGTM, could you revisit the 2 namings I marked?

It is great that we have a task description now, so we can distinguish different download tasks
